src/script/lua_api/l_object.cpp

```diff
@@ -1586,44 +1586,212 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
 	return 1;
 }
 
-// set_sky(self, bgcolor, type, list, clouds = true)
+// set_sky(self, {bgcolor, type, list, clouds = true, 
+//                sun = {true, true, 90, 0, "sun.png"},
+//                moon = {true, -90, 0, "moon.png"},
+//                stars = {true, 200}
+//               })
 int ObjectRef::l_set_sky(lua_State *L)
 {
 	NO_MAP_LOCK_REQUIRED;
 	ObjectRef *ref = checkobject(L, 1);
 	RemotePlayer *player = getplayer(ref);
-	if (player == NULL)
+	if (!player)
 		return 0;
+	if (lua_istable(L, 2)) {
+
+
+		SkyParams sky_params = player->getSkyParams();
+		
+		// just in case we're not supplied a bgcolor
+		sky_params.bgcolor = video::SColor(255,255,255,255);
+
+		lua_getfield(L, 2, "sky_color");
+		if (!lua_isnil(L, -1)) {
+			read_color(L, -1, &sky_params.bgcolor);
+		}
+		lua_pop(L, 1);
+		
+		lua_getfield(L, 2, "type");
+		if (!lua_isnil(L, -1))
+			sky_params.type = luaL_checkstring(L, -1);
+		lua_pop(L, 1);
+		
+		lua_getfield(L, 2, "textures");
+
+		sky_params.params.clear();
+		if (lua_isnil(L, -1)) {
+			sky_params.params.emplace_back("__not_supplied__");
+		} else {
+			lua_pushnil(L);
+			while (lua_next(L, -2) != 0) {
+				// key at index -2 and value at index -1
+				if (lua_isstring(L, -1)) {
+					sky_params.params.emplace_back(readParam<std::string>(L, -1));
+				} else
+					sky_params.params.emplace_back("");
+				// removes value, keeps key for next iteration
+				lua_pop(L, 1);
+			}
+		}
+		lua_pop(L, 1); // Pop the textures table identifier
+
+		// We want to throw an error only when the mod specifies all but
+		// six textures, but not for the custom skybox.
+		if (sky_params.params.size() != 6 && sky_params.type == "skybox")
+			throw LuaError("Skybox expects 6 textures.");
+		if (sky_params.params.size() != 6 && sky_params.type == "custom" &&
+				sky_params.params[0] != "__not_supplied__")
+			throw LuaError("Skybox expects 6 textures.");
+		
+		if (!lua_isnil(L, -1))
+			sky_params.clouds          = getboolfield_default(L, -1,
+				"clouds", true);
+
+		if (sky_params.type != "custom") {
+			getServer(L)->setSky(player, sky_params);
+			lua_pushboolean(L, true);
+			return 1;
+		}
+	
+		if (!lua_isnil(L, -1))
+			sky_params.default_fog     = getboolfield_default(L, -1,
+				"default_fog", false);
+		if (!lua_isnil(L, -1))
+			sky_params.overlay_visible = getboolfield_default(L, -1,
+				"overlay_visible", false);
+
+		lua_getfield(L, 2, "sun"); // Moon related object management
+
+		sky_params.sun.visible      = getboolfield_default(L, -1, 
```

Comment:
Lining up is not practical here because it requires line wrapping; makes it even less readable.